### ISSUE TITLE
[Feature request] Resolve the hostname of containers which are part of a service via DNS service discovery

### ISSUE THREAD

**Issue by kleinsasserm** [moby/moby#34239]
This is a feature request regarding the possibility to resolve the container given hostname from other containers in the same Docker swarm stack. I've already opened an issue at swarmkit about it docker/swarmkit#2325. There are also the details listed.

Now I've looked forward into the code and I guess that I've found the correct place to patch the docker/libnetwork  source with two simple lines of code to implement this feature. I will open a PR for this feature branch.

**Why is this feature needed:**
During the deployment of Docker stack services it would be very useful to automatically have the possibility to resolve the other container ip-addresses of services via their hostnames. The DNS-RR which is automatically created is often not enough.

**Example:**
You would like to deploy a cluster server via Docker swarm with zero configuration overhead. Therefore you will create a service with, for example three replicas. These replicas need to now each other to get the cluster up and running. Due to the dynamic of services in Docker, which is a great thing, you can't predict the ip-adresses of the individual containers and they are a subject of change during a redeployment but you need this because the cluster nodes must reach each other (Raft implementations like Etcd/Redis/...). Given that, there has to be an option to resolve the given hostnames inside the service discovery. See docker/swarmkit#2325 for details. If this would work, you can specify the hostnames (they are predictable and deterministic) inside you cluster configuration files, because every container could reach each other via the deterministic hostnames. Eg --join test_srv.1 test_srv.2 test_srv3. If the containers restarts or the stack is rebuild, there will be three containers with this names. The cluster will bootstrap and continues to work regardless of the given container ip's. 

Details about what and how I did will come with the PR.

**Comment by japrogramer** [comment]
I am able to do this, from within a container. all i have to do is ping service_name or ping image;
for example
`ping postgres`
it stops working randomly tho, see #34240

**Comment by kleinsasserm** [comment]
Yes, pinging the service is not the problem. Have you ever tried to ping the hostname of running containers? See docker/swarmkit#2325. Hostnames of the containers are not registered at DNS, never. Pinging registry_app.1.oc3eudiz86attr7kktjk0jjdl in the example works but the task id is not very useful because it will change every time the container is refreshed.

```
[09:27 mario@ ~/reg]$ docker exec registry_app.1.oc3eudiz86attr7kktjk0jjdl hostname
registry_app-1
[09:28 mario@ ~/reg]$ docker exec registry_app.2.w4mjpo46ikov2xkero552r6if hostname
registry_app-2
[09:28 mario@ ~/reg]$ docker exec registry_app.1.oc3eudiz86attr7kktjk0jjdl ping registry_app-2
ping: bad address 'registry_app-2'
```
It should be (PR):

```
root@b31f5a9da046: # docker service create --network test_net --name test_srv --hostname={{.Service.Name}}-{{.Task.Slot}} --replicas 2 n0r1skcom/echo

root@b31f5a9da046:/go/src/github.com/docker/docker# docker exec test_srv.1.0f9x9a0x46xi7fi81qi57ntyy hostname
test_srv-1
root@b31f5a9da046:/go/src/github.com/docker/docker# docker exec test_srv.2.jttdg37ra8hh3yx7gl2nnff42 hostname
test_srv-2
root@b31f5a9da046:/go/src/github.com/docker/docker# docker exec test_srv.1.0f9x9a0x46xi7fi81qi57ntyy ping test_srv-1
PING test_srv-1 (10.0.0.3): 56 data bytes
64 bytes from 10.0.0.3: icmp_seq=0 ttl=64 time=0.075 ms
64 bytes from 10.0.0.3: icmp_seq=1 ttl=64 time=0.119 ms
```

**Comment by fcrisciani** [comment]
@kleinsasserm, thanks for the PR, here is my point of view.
Hostname is today not used because we have no guarantee to ensure uniqueness of it, so it will be always possible to create it in such a way that will clash between different services.

From a theoretical point of view, a service represent a way to define an application for which you don't have a need to point to a single task. Tasks are ephemeral, they are shuffled around nodes, they can crash, they can scale up and down. The advantage is that the scheduler will do its best to match the intent of the service itself in an autonomous way. The objective is to make the user of the application believe the service is always healthy, no matter from the scale or nodes failures.

If you have an application that requires to uniquely point to a container, you are going beyond the idea of the service itself where all the tasks are the same, so maybe the solution is to run single containers with docker run and network aliases to be able to uniquely identify them.

Instead if you want to pursue the way of the service, we are providing the possibility to resolve all the tasks that are active at a given time as: `nslookup tasks.<service_name>`. This will return the list of the IPs of the tasks that are available.
In your case you can consider to add a script in your docker image that wait for this list to match the number of replicas that you expect and then pass these IPs to your application to initialize. (Keep in mind that the IP can change during the execution)

Another thing to keep in mind once you use a template like `{{.Service.Name}}-{{.Task.Slot}}` is there is no guarantee that if you select 3 replicas the Task.Slot would always be only 1,2 and 3.

If you try to create a service with 2 replicas, then scale it up to 20 and then scale it down to 2 again you will see something like this:
```
docker service ps test_srv                                      
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE               ERROR               PORTS                                                                                                                       
xn4g83otogra        test_srv.2          alpine:latest       c9665c1c97a1        Running             Running about an hour ago                                                                                                                                                   
ii7hoziq9n0x        test_srv.11         alpine:latest       c9665c1c97a1        Running             Running 14 seconds ago   
```
As you can see I have tasks 2 and tasks 11 that are alive at the end.

We are looking into creating a templated task alias, but still the behavior that I was describing above will be the same so has to be clearly understood while using services

**Comment by kleinsasserm** [comment]
Yes, I know about the scaling problem, especially about the scale down problem - that is clearly true, no objection at this point. But not all services are meant to be scaled up and down because they need a more deterministic setup.

Of course in a perfect world all services are stateless and scaleable, but unfortunately that is not true in real live environment with software that you cannot influence. For the other readers, we discuss this topic at docker/libnetwork#1855. Many thanks to all commentators - its fun to discuss!

**Comment by chrisbecke** [comment]
What. Still nothing. I mean the following looks simple, obvious and wrong:

```yaml
services:
  minio:
    image: minio/minio
    hostname: minio{{.Task.Slot}}
    command: server http://minio{1..4}:9000
    deploy:
      replicas: 4
```

**Comment by ajozz13** [comment]
quick question...is it possible to define what the hostname entry would be on a service?
currently when deploying in swarm node the hostname for the service is set to:
service-name_app ...this name is in the dns entry.  Unfortunately the _ underscore option makes it impossible to add that name as part of a URL  (apache)  in the form http://service-name_app/  
can we modify this dns entry en any manner?  I could not see an anwer about adding the hostname as an option, is there a way to change that DNS entry from the docker-compose file?

**Comment by chrisbecke** [comment]
Addressed by #39204 looks like

**Comment by achempion** [comment]
It looks like it wouldn't work for containers running in a `global` mode.

> Docker version 19.03.12, build 48a66213fe

**Comment by thaJeztah** [comment]
https://github.com/moby/moby/pull/39204 is not yet in a release, only in the master branch

**Comment by achempion** [comment]
@thaJeztah thank you for pointing out. 

This comment states that with 20.03 it will work but only for replicated containers:
https://github.com/moby/moby/pull/39204#issuecomment-597801026.
